File: app/src/main/java/com/perflyst/twire/tasks/GetVisualElementsTask.java
package com.perflyst.twire.tasks;

import android.util.Log;

import com.perflyst.twire.activities.main.LazyFetchingActivity;
import com.perflyst.twire.utils.Execute;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

public class GetVisualElementsTask<T> implements Callable<List<T>> {
    private final LazyFetchingActivity<T> mLazyActivity;
    private final String LOG_TAG = getClass().getSimpleName();

    public GetVisualElementsTask(LazyFetchingActivity<T> mLazyActivity) {
        this.mLazyActivity = mLazyActivity;
    }

    public final List<T> call() {
        final List<T> resultList = new ArrayList<>();

        try {
            resultList.addAll(mLazyActivity.getVisualElements());
        } catch (Exception e) {
            e.printStackTrace();
        }

        Execute.ui(() -> {
            if (resultList.isEmpty()) {
                Log.i(LOG_TAG, "ADDING 0 VISUAL ELEMENTS");
                mLazyActivity.notifyUserNoElementsAdded();
            }

            mLazyActivity.addToAdapter(resultList);
            mLazyActivity.stopProgress();
            mLazyActivity.stopRefreshing();
        });

        return resultList;
    }
}

File: app/src/main/java/com/perflyst/twire/tasks/GetVisualElementsTask.kt
package com.perflyst.twire.tasks

import com.perflyst.twire.activities.main.LazyFetchingActivity
import com.perflyst.twire.utils.Execute
import timber.log.Timber
import java.util.concurrent.Callable

class GetVisualElementsTask<T>(private val mLazyActivity: LazyFetchingActivity<T>) :
    Callable<MutableList<T>> {
    override fun call(): MutableList<T> {
        val resultList: MutableList<T> = ArrayList()

        try {
            resultList.addAll(mLazyActivity.visualElements)
        } catch (_: InterruptedException) {
            return resultList
        } catch (e: Exception) {
            Timber.e(e)
        }

        Execute.ui {
            if (resultList.isEmpty()) {
                Timber.i("ADDING 0 VISUAL ELEMENTS")
                mLazyActivity.notifyUserNoElementsAdded()
            }
            mLazyActivity.addToAdapter(resultList)
            mLazyActivity.stopProgress()
            mLazyActivity.stopRefreshing()
        }

        return resultList
    }
}
